akka-projection-grpc/src/main/scala/akka/projection/grpc/internal/FilterStage.scala

```diff
@@ -110,22 +161,11 @@ import akka.stream.stage.StageLogging
             val env = grab(in2)
             val pid = env.persistenceId
 
-            // FIXME this logic is too fuzzy
-            val exclude =
-              if (includePersistenceIds.isEmpty && excludePersistenceIds.isEmpty)
-                false
-              else if (includePersistenceIds(pid))
-                false
-              else if (!excludePersistenceIds(pid))
-                false
-              else
-                true
-
-            if (exclude) {
+            if (filter.matches(pid)) {
+              push(out2, env)
+            } else {
```

Comment:
Don't we still need to push empty/filtered envelope?

Reply:
Not when we exclude all events for an persistenceId. The purpose of the filtered marker is to still store an offset for it on the consumer side, so that it can accept next seqNr for that persistenceId.

Here, if we later change the filter to include the persistenceId it will have to replay previous events first, and that is what I'm working on in an upcoming PR.